app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Preferences/SharedPreferencesGrouper.java

```diff
@@ -74,21 +78,88 @@ public SharedPreferences getSharedPreference(SPType key){
+        if(listenerList == null || !listenerList.contains(listener)) {
             Log.e(TAG, "Listener not found for "+key);
             return;
         }
         sp.unregisterOnSharedPreferenceChangeListener(listener);
-        this.spMap.remove(key);
+        listenerList.remove(listener);
+        // If listener list isn't empty, write in the new reduced list, otherwise just delete the whole key
+        if (!listenerList.isEmpty()) {
+            this.spMap.put(key, listenerList);
+        } else {
+            this.spMap.remove(key);
+        }
         Log.i(TAG, "Listener removed for "+key);
     }
+
+    /**
+     * Removes all listeners for a particular {@link SPType} {@code key}
+     * @param key the key for which to remove all listeners
+     */
+    public void removeAllListeners(SPType key){
+        SharedPreferences sp = this.getSharedPreference(key);
+        if(sp == null) {
+            Log.e(TAG, "SharedPreferences not found for "+key);
+            return;
+        }
+        Set<SharedPreferences.OnSharedPreferenceChangeListener> listenerList = this.spMap.get(key);
+        if(listenerList == null) {
+            Log.e(TAG, "Listener(s) not found for "+key);
+            return;
+        }
+        // Unregister all listeners
+        for (SharedPreferences.OnSharedPreferenceChangeListener listener : listenerList) {
+            sp.unregisterOnSharedPreferenceChangeListener(listener);
+        }
+        // Delete them all from the map
+        this.spMap.remove(key);
+        Log.i(TAG, "Listener(s) removed for "+key);
+    }
+
+    /**
+     * Searches for the passed-in {@code listener} and removes it if found. The listener
+     * <b>must have been</b> already added via {@link #setListener(SharedPreferences.OnSharedPreferenceChangeListener, SPType)}
+     * @param listener the listener to add
+     */
+    public void removeListener(@NonNull SharedPreferences.OnSharedPreferenceChangeListener listener){
+        for (SPType potentialMatch : spMap.keySet()) {
+            SharedPreferences sp = this.getSharedPreference(potentialMatch);
+            if (sp == null) {
+                Log.e(TAG, "SharedPreferences not found for " + potentialMatch);
+                continue;
+            }
+            Set<SharedPreferences.OnSharedPreferenceChangeListener> listenerList = this.spMap.get(potentialMatch);
+            if (listener == null || !listenerList.contains(listener)) {
+                Log.e(TAG, "Listener not found for " + potentialMatch);
+                continue;
+            }
+            // Now the easy part, when we know which SharedPref owns the listener and it's in the
+            // registered listenerList, just remove it from the list and cleanup the key if the list
+            // remains empty
+            listenerList.remove(listener);
+            sp.unregisterOnSharedPreferenceChangeListener(listener);
```

Comment:
what i'm saying is, it won't matter if we remove from list first then unregister, or unregister followed by removing from the list because there is no real way of knowing whether the unregistration happened yet, or if anything was unregistered at all.

resolving.